### PR TITLE
feat: Add worker configs var and handler

### PR DESCRIPTION
**Pull request by dkoshkin**
This adds code to set worker overrides to clusterConfig, here is an example of how to use it:

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: <NAME>
spec:
    topology:
      class: docker-quick-start
      controlPlane:
        metadata: {}
        replicas: 1
      variables:
      - name: clusterConfig
        value:
          addons:
            cni:
              provider: calico
            nfd: {}
          controlPlane:
            docker:
              customImage: ghcr.io/mesosphere/kind-node:v1.2.3-cp
          docker: {}
      - name: workerConfig
        value:
          docker:
            customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker
      version: v1.27.5
      workers:
        machineDeployments:
        - class: default-worker
          metadata: {}
          name: md-0
          replicas: 1
        - class: default-worker
          metadata: {}
          name: md-1
          replicas: 1
          variables:
            overrides:
            - name: workerConfig
              value:
                docker:
                  customImage: ghcr.io/mesosphere/kind-node:v1.2.3-custom
```

### CODE REVIEW

File: pkg/handlers/docker/mutation/customimage/inject.go [hunk 1]
```diff
@@ -149,6 +159,10 @@ func (h *customImagePatchHandler) Mutate(
 		return err
 	}
 
+	if h.variableName == workerconfig.MetaVariableName {
+		return nil
+	}
+
```
Comment (dkoshkin) [inline pkg/handlers/docker/mutation/customimage/inject.go:165]: Not a fan of this

**Comment by dkoshkin** [comment]
@supershal this may be useful to look at for https://github.com/d2iq-labs/capi-runtime-extensions/issues/199

File: pkg/handlers/docker/mutation/customimage/inject.go (hunk 1 above)
Comment (jimmidyson) [reply pkg/handlers/docker/mutation/customimage/inject.go:165]: This feels like they should be separate handlers, one for control plane and one for workers. WDYT?

File: docs/content/customization/docker/custom-image.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: Cluster
+metadata:
+  name: <NAME>
+spec:
+  topology:
+    variables:
+      - name: clusterConfig
+        values:
```
Comment (supershal) [inline docs/content/customization/docker/custom-image.md:23]:
```suggestion
        value:
```

File: api/v1alpha1/workerconfig_types.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+func (GenericWorkerConfig) VariableSchema() clusterv1.VariableSchema {
+	return clusterv1.VariableSchema{
+		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+			Description: "Worker Node configuration",
+			Type:        "object",
+			Properties:  map[string]clusterv1.JSONSchemaProps{},
+		},
+	}
+}
```
Comment (supershal) [inline api/v1alpha1/workerconfig_types.go:75]:
```suggestion
}

func init() {
	SchemeBuilder.Register(&WorkerConfig{})
}
```

File: docs/content/customization/docker/custom-image.md (hunk 2 above)
Comment (dkoshkin) [reply docs/content/customization/docker/custom-image.md:23]: Thanks, its actually `values:` here, but changed fixed it in the other doc.

Comment (supershal) [reply docs/content/customization/docker/custom-image.md:23]: may be I am looking at wrong place. I think https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/cluster_types.go#L327 is related to this object?? 🤔

Comment (dkoshkin) [reply docs/content/customization/docker/custom-image.md]: 🤦 yep you're right! I was looking at a different example in the docs which had `values`

File: pkg/handlers/docker/mutation/customimage/inject_worker.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+		h.variableFieldPath,
+		"variableValue",
+		customImageVar,
+	)
+
+	return patches.Generate(
+		obj,
+		vars,
+		&holderRef,
+		selectors.InfrastructureWorkerMachineTemplates("v1beta1", "DockerMachineTemplate"),
```
Comment (dkoshkin) [inline pkg/handlers/docker/mutation/customimage/inject_worker.go:116]: :( I don't really understand how this selector works, why is the worker patch getting called with control-plane vars (also why the unit tests are failing):
```
  topology:
    class: docker-quick-start
    controlPlane:
      metadata: {}
      replicas: 1
    variables:
      - name: clusterConfig
        value:
          addons:
            cni:
              provider: calico
            nfd: {}
          docker: {}
          controlPlane:
            docker:
              customImage: ghcr.io/mesosphere/kind-node:v1.2.3-cp
          workers:
            docker:
              customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker
    version: v1.27.5
    workers:
      machineDeployments:
        - class: default-worker
          name: md-0
          replicas: 1
        - class: default-worker
          name: md-1
          replicas: 1
          variables:
            overrides:
              - name: workerConfig
                value:
                  docker:
                    customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker-md-1
```

```
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  metadata:
    labels:
      topology.cluster.x-k8s.io/deployment-name: md-1
    name: docker-quick-start-md-1-infra-6fdjg
  spec:
    template:
      spec:
        customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker-md-1
        extraMounts:
        - containerPath: /var/run/docker.sock
          hostPath: /var/run/docker.sock
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  metadata:
    labels:
      topology.cluster.x-k8s.io/deployment-name: md-0
    name: docker-quick-start-md-0-infra-9ckdc
  spec:
    template:
      spec:
        customImage: ghcr.io/mesosphere/kind-node:v1.27.5
        extraMounts:
        - containerPath: /var/run/docker.sock
          hostPath: /var/run/docker.sock
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  metadata:
    name: docker-quick-start-control-plane-mq6p5
  spec:
    template:
      spec:
        customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker
        extraMounts:
        - containerPath: /var/run/docker.sock
          hostPath: /var/run/docker.sock
```

**Comment by dkoshkin** [comment]
Ok moving `workerConfig` to a separate variable really simplified it, working as expected:
```
    topology:
      class: docker-quick-start
      controlPlane:
        metadata: {}
        replicas: 1
      variables:
      - name: clusterConfig
        value:
          addons:
            cni:
              provider: calico
            nfd: {}
          controlPlane:
            docker:
              customImage: ghcr.io/mesosphere/kind-node:v1.2.3-cp
          docker: {}
      - name: workerConfig
        value:
          docker:
            customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker
      version: v1.27.5
      workers:
        machineDeployments:
        - class: default-worker
          metadata: {}
          name: md-0
          replicas: 1
        - class: default-worker
          metadata: {}
          name: md-1
          replicas: 1
          variables:
            overrides:
            - name: workerConfig
              value:
                docker:
                  customImage: ghcr.io/mesosphere/kind-node:v1.2.3-custom
```

```
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  metadata:
    name: docker-quick-start-control-plane-pqkd9
    namespace: default
  spec:
    template:
      spec:
        customImage: ghcr.io/mesosphere/kind-node:v1.2.3-cp
        extraMounts:
        - containerPath: /var/run/docker.sock
          hostPath: /var/run/docker.sock
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  metadata:
    name: docker-quick-start-md-0-infra-9526d
    namespace: default
  spec:
    template:
      spec:
        customImage: ghcr.io/mesosphere/kind-node:v1.2.3-worker
        extraMounts:
        - containerPath: /var/run/docker.sock
          hostPath: /var/run/docker.sock
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  metadata:
    name: docker-quick-start-md-1-infra-b5cjx
    namespace: default
  spec:
    template:
      spec:
        customImage: ghcr.io/mesosphere/kind-node:v1.2.3-custom
        extraMounts:
        - containerPath: /var/run/docker.sock
          hostPath: /var/run/docker.sock

File: pkg/handlers/docker/mutation/customimage/inject.go (hunk 1 above)
Comment (dkoshkin) [reply pkg/handlers/docker/mutation/customimage/inject.go]: 👍 yep split into multiple handlers

**Comment by jimmidyson** [comment]
Nice! I won't be able to review or try this out until Monday so help from @supershal or @faiq would be great to get this merged :pray:

**Comment by supershal** [comment]
> Nice! I won't be able to review or try this out until Monday so help from @supershal or @faiq would be great to get this merged 🙏

me and @dkoshkin just go over the PR code changes this morning. changes looks good. I will test it locally before approving it today.

**Review by supershal — APPROVED** [review]

**Comment by supershal** [comment]
@dkoshkin should we remove [region/variables.go](https://github.com/d2iq-labs/capi-runtime-extensions/blob/main/pkg/handlers/aws/mutation/region/variables.go) and its test file with this change?

**Comment by dkoshkin** [comment]
> @dkoshkin should we remove [region/variables.go](https://github.com/d2iq-labs/capi-runtime-extensions/blob/main/pkg/handlers/aws/mutation/region/variables.go) and its test file with this change?

I don't think so @supershal , that configuration is still valid. Are you seeing something in this PR that made you think that?

**Comment by supershal** [comment]
the `awsRegionVariableHandler` and `customImageVariableHandler` are not registered anymore. their unit testing is covered by test in the `clusterconfig` package. https://github.com/d2iq-labs/capi-runtime-extensions/blob/main/pkg/handlers/aws/clusterconfig/variables_test.go

**Comment by dkoshkin** [comment]
> the `awsRegionVariableHandler` and `customImageVariableHandler` are not registered anymore. their unit testing is covered by test in the `clusterconfig` package. https://github.com/d2iq-labs/capi-runtime-extensions/blob/main/pkg/handlers/aws/clusterconfig/variables_test.go

Lets do that in a separate PR, we could do this cleanup in all handlers not just these two.

**Pull request merged**